Sources/FluentKit/Properties/Parent.swift

```diff
@@ -1,69 +1,120 @@
+        var description: String {
+            switch self {
+            case let .loaded(model): return model.description
+            case .notLoaded: return "notLoaded"
             }
-            return parent
         }
     }
-
 }
 
 extension Parent: FieldRepresentable {
-    public var field: Field<To.IDValue> {
-        return self.$id
-    }
+    public var field: Field<To.IDValue> { self.$id }
 }
 
 extension Parent: AnyProperty {
     func encode(to encoder: Encoder) throws {
         var container = encoder.singleValueContainer()
-        if let parent = self.eagerLoadedValue {
+
+        if case let .loaded(parent) = self.eagerLoadedValue {
             try container.encode(parent)
         } else {
-            try container.encode([
-                To.key(for: \._$id): self.id
-            ])
+            try container.encode([self.$id.key: self.id])
         }
     }
 
     func decode(from decoder: Decoder) throws {
         let container = try decoder.container(keyedBy: _ModelCodingKey.self)
-        try self.$id.decode(from: container.superDecoder(forKey: .string(To.key(for: \._$id))))
+        try self.$id.decode(from: container.superDecoder(forKey: .string(self.$id.key)))
         // TODO: allow for nested decoding
     }
 }
 
 extension Parent: AnyField { }
+
+
+// MARK: - Specialization
+
+extension Parent where To: Model {
+    public convenience init(key: String) {
+        self.init(id: key, eagerLoadRequest: SubqueryEagerLoad(key: key))
+    }
+
+    public func query(on database: Database) -> QueryBuilder<To> {
+        return To.query(on: database).filter(\._$id == self.id)
+    }
+
+    public func get(on database: Database) -> EventLoopFuture<To> {
+        return self.query(on: database).first().flatMapThrowing { parent in
+            guard let parent = parent else {
+                throw FluentError.missingParent
+            }
+            return parent
+        }
+    }
+}
+
+extension Parent where To: OptionalType, To.Wrapped: Model, To.IDValue == To.Wrapped.IDValue? {
+    public convenience init(key: String) {
+        self.init(id: key, eagerLoadRequest: SubqueryEagerLoad(key: key))
+    }
+
+    public func query(on database: Database) -> QueryBuilder<To.Wrapped> {
+        return To.Wrapped.query(on: database).filter(\To.Wrapped._$id == self.id)
+    }
+
+    public func get(on database: Database) -> EventLoopFuture<To> {
+        return self.query(on: database).first().flatMapThrowing { parent in parent as! To }
+    }
+}
+
+
+// MARK: - Optionals
+
+public typealias GenericModel = ModelIdentifiable & Codable & CustomStringConvertible
+
+extension Optional: CustomStringConvertible where Wrapped: CustomStringConvertible {
+    public var description: String {
+        return self?.description ?? "nil"
+    }
+}
+
+extension Optional: ModelIdentifiable where Wrapped: ModelIdentifiable {
+    public typealias IDValue = Wrapped.IDValue?
+
+    public var id: Wrapped.IDValue?? {
```

Comment:
Do we need a double optional here?

Reply:
Yes. The reason is because the `Model` type which originally had the ID information requires that the type of the `.id` property is the optional of `IDValue`. I copied this over to the `ModelIdentifiable` protocol to prevent breaking changes. Since the `IDValue` of `Optional<ModelIdentifiable>` needs to be `Optional<Wrapped.IDValue>`, the optional version of that which we use for the `.id` property is `Optional<Optional<Wrapped.IDValue>>`

Reply:
I see the reasoning here now. If this is not `??` then the Parent's `$id` field will be a non-optional. This seems like code smell to me. In other words, this protocol taken by itself is confusing with the double optional. It only makes sense in the context of `Parent`. This makes me think that the protocol should be called `ParentRelatable` or something so that the correlation is clearer.